src/Microsoft.ML.ModelBuilder.SearchSpace/Attribute/RangeAttribute.cs

```diff
@@ -0,0 +1,47 @@
+﻿// <copyright file="RangeAttribute.cs" company="Microsoft">
+// Copyright (c) Microsoft. All rights reserved.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using Microsoft.ML.ModelBuilder.SearchSpace.Option;
+
+namespace Microsoft.ML.ModelBuilder.SearchSpace
+{
+    [AttributeUsage(AttributeTargets.Property | AttributeTargets.Field, Inherited = false, AllowMultiple = false)]
+    public class RangeAttribute : Attribute
+    {
+        public RangeAttribute(double min, double max, bool logBase = false)
+        {
+            this.Option = new UniformDoubleOption(min, max, logBase);
+        }
+
+        public RangeAttribute(double min, double max, double init, bool logBase = false)
+        {
+            this.Option = new UniformDoubleOption(min, max, logBase, init);
+        }
```

Comment:
Can't you combine all of these into a single constructor with `init` defaulting to `null`? The same way you have the constructor for `UniformDoubleOption` itself. 

This applies to all the other types as well.

Reply:
Attribute constructor can't accept nullable type so it has to be separate...

Reply:
What about using `Double.NaN` and `Float.NaN` instead of using a nullable type? That might be easier overall and should work here.

Reply:
Would it be confusing? `Double.NaN` has its own meaning (divided  by 0 for example) and would it makes user feel like the init value is `Double.NaN` instead?

Reply:
NaN just means not a number. Which is true in this case as well. When text files are read in, if a column is a float/double column and the column is empty, we initialize the value to NaN. So I think using NaN for stuff like this is pretty common.

@eerhardt Do you have any thoughts for situations like these?

Reply:
Different repos do different things here and it really is just dependent on if we expect `NaN` could be a valid input for some other context.

In many contexts, `NaN` is roughly equivalent to `null` and represents an error case or an unused/uninitialized value and so can be used to avoid needing to track some separate overload or use `Nullable<T>` or track a separate `bool`, etc. WPF actually uses it for this purpose.

In the case of a `RangeAttribute`, I would consider `NaN` to be like `null`. These already have to be "constant" and so producing it as the result of some constant expression is unlikely.

Reply:
There's a chance that people might think an option would use `Double.NaN` as the init value to start instead of starting from random, which is what it actually means since `Double.NaN` is an actual constant value in CSharp. To avoid that risk I would prefer to go through the dumb way and provides two api for creating an option with or without init value.

Reply:
You can't define `NaN` as some start value. It is not comparable to any values and is not greater, lesser, or equal to any other value including `NaN`.

Reply:
Basically `NaN` is `NaN` and any operation involving `NaN` is generally meant to produce `NaN` (generally the same `NaN` that was input).

The only thing you can really do with `NaN` is special-case it with `double.IsNaN(x)` and so it generally fills the same role as `null` which is to indicate "nothing" or "failure".

Reply:
That's why I say it's confusing. Passing a `NaN` as a parameter doesn't make any sense because it's not comparable to any other value, so we should avoid having a constructor that accepts `NaN` as input. Because it's not self-explainable enough and people will be confused at the first glance unless they seek into code or comment.

Moreover, in machine learning, `NaN` almost always means something bad is going on (overflow, div by 0,,,), that's also why we should avoid it, not to say using it as default value.

Reply:
I wouldn't agree. `NaN` is commonly used to represent a default/uninitialized value (including in ML spaces since it saves space and will be more efficient than tracking initialization or lack of value separately) as well.

Its just one of the many weird floating-point quirks users may need to be familiar with.

Reply:
A lot of data processing using `NaN` instead of `null` to replace missing value and that makes sense, but it's mostly like a trade-off since it would be more efficient in some operations (like mask). And such trade-off is not necessary in this case, right? 

`NaN` is not a solid input based on common knowledge, and that's what I mostly worried about. Using `NaN` instead of `null` to represent a missing value in user data is one story, but using it as a default parameter is another one. The first is fine because we are just using a flag to replace invalid data, but the second is not because we are expecting an invalid (not a number) input and translate it into a valid one (null).

Reply:
> The first is fine because we are just using NaN as a flag to replace invalid data, but the second is not because we are expecting an invalid (not a number) input and translate it into a valid one (null).

Its the same here. `NaN` is just a flag saying "no data was provided". That then gets treated as a relevant default.

It's done in WPF and other libraries. It's the same as using `null` or `-1` as a placeholder value and is often done simply because that represents an otherwise "invalid input".

Most APIs reject `null` and will throw if you pass it in. Others, like say `string.Format` treat it as a "default" meaning "replace me with the correct default" (often `format: "G"`). The same applies to buffer APIs where we can't necessarily use `0` because `0` is a valid value and could mean "no buffering" while `-1` on the other hand can never be valid and so the options are either `throw` or treat it like a sentinel value to mean "replace me with the correct default".

Throughout the BCL, we use a mix of `sentinel values` (`null`, `NaN`, `negatives`, `MaxValue`, etc) and `nullables` (such as `float?` or `int?`) depending on the scenario and intended use-case (often preferring sentinels in the case where perf matters and nullables where clarity matters or where there is no available sentinel).

I'm not necessarily saying that using `NaN` is the best choice here (its an attribute, perf doesn't really matter, etc). Just trying to clarify that we shouldn't be shying away from `NaN` just because its `NaN`.

Reply:
It looks like typically we have been using null for public api (https://github.com/dotnet/machinelearning/blob/510f0112d4fbb4d3ee233b9ca95c83fae1f9da91/src/Microsoft.ML.LightGbm/LightGbmCatalog.cs#L42). I'm fine either way, the only reason I suggested using `double.NaN` is that you could avoid having 2 constructors.

You said that the attribute constructors can't accept nullable types. How come?

Reply:
check https://docs.microsoft.com/en-us/previous-versions/visualstudio/visual-studio-2008/7kk5syz6(v=vs.90)?redirectedfrom=MSDN

Reply:
Attributes only allow a limited subset of values to be used/encoded (generally things that can be constant in IL). Nullable (and generics) aren't one of those things.